src/ring_theory/jacobson.lean

```diff
@@ -252,4 +259,156 @@ end
+    quotient_map (I.comap φ') ϕ.to_map
+    (le_of_eq (trans (comap_comap φ ϕ'.to_map) (hcomm ▸ (comap_comap ϕ.to_map φ').symm))),
+  have : ((I.comap ϕ'.to_map).comap φ).is_maximal := (by rw [comap_comap, hcomm, ← comap_comap] :
+    ((I.comap φ').comap ϕ.to_map) = ((I.comap ϕ'.to_map).comap φ)) ▸
+      ((is_maximal_iff_is_maximal_disjoint ϕ _).1
+      (is_maximal_comap_of_is_integral_of_is_maximal' φ' hφ' I hI)).left,
+  rw is_maximal_iff_bot_quotient_is_maximal at this ⊢,
+  refine is_maximal_of_is_integral_of_is_maximal_comap' f _ ⊥
+    ((eq_bot_iff.2 (comap_bot_le_of_injective f quotient_map_injective)).symm ▸ this),
+  exact is_integral_tower_bot_of_is_integral' f g quotient_map_injective
+    ((comp_quotient_map_eq_of_comp_eq hcomm I).symm ▸
+    (ring_hom.is_integral_trans (is_integral_of_surjective'
+      (localization_map.surjective_quotient_map_of_maximal_of_localization
+      (by rwa [comap_comap, hcomm, is_maximal_iff_bot_quotient_is_maximal])))
+      (is_integral_quotient_of_is_integral' hφ'))),
+end
+
+/-- Used to bootstrap the proof of `is_jacobson_polynomial_iff_is_jacobson`.
+  That theorem is more general and should be used instead of this one -/
+private lemma is_jacobson_polynomial_of_domain (R : Type*) [integral_domain R] [hR : is_jacobson R]
+  (P : ideal (polynomial R)) [P.is_prime] (hP : ∀ (x : R), C x ∈ P → x = 0) : P.jacobson = P :=
+begin
+  by_cases hP : (P = ⊥),
+  { exact hP.symm ▸ jacobson_bot_polynomial_of_jacobson_bot (hR ⊥ radical_bot_of_integral_domain) },
+  { rw jacobson_eq_iff_jacobson_quotient_eq_bot,
+    let P' : ideal R := P.comap C,
+    have hP'_inj : function.injective (quotient.mk P') := (quotient.mk P').injective_iff.2
+      (λ x hx, by rwa [quotient.eq_zero_iff_mem, (by rwa eq_bot_iff : P' = ⊥)] at hx),
+    haveI : P'.is_prime := comap_is_prime C P,
+    have : ∃ (p : polynomial R) (hp : p ∈ P), p ≠ 0,
+    { contrapose! hP,
+      exact eq_bot_iff.2 (λ x hx, (hP x hx).symm ▸ (ideal.zero_mem ⊥)) },
+    obtain ⟨p, hp, hp0⟩ := this,
+    have hp0 : (p.map (quotient.mk P')) ≠ 0 :=
+      λ hp0', hp0 $ map_injective (quotient.mk P') hP'_inj (by simpa using hp0'),
+    let φ : P'.quotient →+* P.quotient := quotient_map P C le_rfl,
+    have hφ' : φ.comp (quotient.mk P') = (quotient.mk P).comp C := rfl,
+    have hφ : function.injective φ := quotient_map_injective,
+    let M : submonoid P'.quotient := submonoid.powers (p.map (quotient.mk P')).leading_coeff,
+    let ϕ : localization_map M (localization M) := localization.of M,
+    let ϕ' : localization_map (M.map ↑φ) (localization (M.map ↑φ)) := localization.of (M.map ↑φ),
+    let φ' : (localization M) →+* (localization (M.map ↑φ)) :=
+      (ϕ.map (M.mem_map_of_mem (φ : P'.quotient →* P.quotient)) ϕ'),
+    refine jacobson_bot_of_integral_localization φ hφ (p.map (quotient.mk P')).leading_coeff
+       (λ hx, hp0 (leading_coeff_eq_zero.1 hx)) ϕ ϕ' _,
+    suffices : φ'.is_integral_elem (ϕ'.to_map ((quotient.mk P) X)),
```

Comment:
I think this might be good to make into a sublemma, but I'm not sure if it should be generalized in some way? This is essentially saying that if the image of X is integral then the whole extension is integral, since it is a generator for the set of polynomials, but I'm not sure lean has a great way to express that, without becoming very cumbersome?

Reply:
I've ended up creating a generalized version of `is_integral_of_mem_closure` that allows arbitrary sets, after realizing I was taking the constants for granted in my original comment. Because the integral closure forms a subring and not an ideal, we have to explicitly include all the constant elements as well as X in order to make the subring closure of the set be `top`, otherwise only the ideal closure will be `top`, and so we need to pass in the arbitrarily large set of constants and not just X.

I haven't privated the original two variable version since it has simpler arguments in the case that you really only need two elements in the set.